File: moqt/internal/message/fetch.go
package message

import "io"

type FetchMessage struct {
	BroadcastPath string
	TrackName     string
	Priority      uint8
	GroupSequence uint64
}

func (f FetchMessage) Len() int {
	var l int
	l += StringLen(f.BroadcastPath)
	l += StringLen(f.TrackName)
	l += 1 // Priority (uint8)
	l += VarintLen(f.GroupSequence)
	return l
}

func (f FetchMessage) Encode(w io.Writer) error {
	msgLen := f.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))
	b, _ = WriteMessageLength(b, uint64(msgLen))
	b, _ = WriteVarint(b, uint64(len(f.BroadcastPath)))
	b = append(b, f.BroadcastPath...)
	b, _ = WriteVarint(b, uint64(len(f.TrackName)))
	b = append(b, f.TrackName...)
	b = append(b, f.Priority)
	b, _ = WriteVarint(b, f.GroupSequence)
	_, err := w.Write(b)
	return err
}

func (f *FetchMessage) Decode(src io.Reader) error {
	size, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	if size > MaxMessageAllocationSize {
		return ErrMessageTooLarge
	}
	b := make([]byte, size)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	str, n, err := ReadString(b)
	if err != nil {
		return err
	}
	f.BroadcastPath = str
	b = b[n:]

	str, n, err = ReadString(b)
	if err != nil {
		return err
	}
	f.TrackName = str
	b = b[n:]

	if len(b) < 1 {
		return ErrMessageTooShort
	}
	f.Priority = b[0]
	b = b[1:]

	num, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	f.GroupSequence = num
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: moq-web/src/internal/message/fetch.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import { MessageDecoder, MessageEncoder, readFull, readVarint } from "./message.ts";

export interface FetchMessageInit {
	broadcastPath?: string;
	trackName?: string;
	priority?: number;
	groupSequence?: number;
}

export class FetchMessage {
	broadcastPath: string;
	trackName: string;
	priority: number;
	groupSequence: number;

	constructor(init: FetchMessageInit = {}) {
		this.broadcastPath = init.broadcastPath ?? "";
		this.trackName = init.trackName ?? "";
		this.priority = init.priority ?? 0;
		this.groupSequence = init.groupSequence ?? 0;
	}

	async encode(w: Writer): Promise<Error | undefined> {
		return MessageEncoder.encode(w, (e) => {
			e.string(this.broadcastPath);
			e.string(this.trackName);
			e.uint8(this.priority);
			e.varint(this.groupSequence);
		});
	}

	async decode(r: Reader): Promise<Error | undefined> {
		let err: Error | undefined;

		let msgLen: number;
		[msgLen, , err] = await readVarint(r);
		if (err) return err;

		const buf = new Uint8Array(msgLen);
		[, err] = await readFull(r, buf);
		if (err) return err;

		const d = new MessageDecoder(buf);

		this.broadcastPath = d.string();
		this.trackName = d.string();
		this.priority = d.uint8();
		this.groupSequence = d.varint();

		if (!d.eof()) {
			return new Error("message length mismatch");
		}

		return undefined;
	}
}
